data/mongodb/v0/main.go

```diff
@@ -0,0 +1,144 @@
+//go:generate compogen readme ./config ./README.mdx
+package mongodb
+
+import (
+	"context"
+	_ "embed"
+	"fmt"
+	"sync"
+
+	"github.com/instill-ai/component/base"
+	"github.com/instill-ai/x/errmsg"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+const (
+	TaskInsert         = "TASK_INSERT"
+	TaskFind           = "TASK_FIND"
+	TaskUpdate         = "TASK_UPDATE"
+	TaskDelete         = "TASK_DELETE"
+	TaskDropCollection = "TASK_DROP_COLLECTION"
+	TaskDropDatabase   = "TASK_DROP_DATABASE"
+)
+
+//go:embed config/definition.json
+var definitionJSON []byte
+
+//go:embed config/setup.json
+var setupJSON []byte
+
+//go:embed config/tasks.json
+var tasksJSON []byte
+
+var once sync.Once
+var comp *component
+
+type component struct {
+	base.Component
+}
+
+type MongoCollectionClient interface {
+	Find(ctx context.Context, filter interface{}, opts ...*options.FindOptions) (cur *mongo.Cursor, err error)
+	FindOne(ctx context.Context, filter interface{}, opts ...*options.FindOneOptions) *mongo.SingleResult
+	InsertOne(ctx context.Context, document interface{}, opts ...*options.InsertOneOptions) (*mongo.InsertOneResult, error)
+	UpdateMany(ctx context.Context, filter interface{}, update interface{}, opts ...*options.UpdateOptions) (*mongo.UpdateResult, error)
+	DeleteMany(ctx context.Context, filter interface{}, opts ...*options.DeleteOptions) (*mongo.DeleteResult, error)
+	Drop(ctx context.Context) error
+}
+
+type MongoDBClient interface {
+	Drop(ctx context.Context) error
+}
```

Comment:
question
Is there a concern that you divide MongoCollectionClient & MongoDBClient?
They seem to be under MongoDBClient.
And, this DB does not mean database but it is the name of MongoDB.

e.g. If the MongoDB name with Mongo, then it could be like
```
type Mongo interface {
  Find()
  FindOne()
...
  DropDatabase()
  CreateDatabase()
...
}
```

Reply:
@chuang8511 

This is the part where I would like your suggestions the most, the reason why I divide it into two structs is so that it could be use for mock input for component test as well.

The structure of mongodb go sdk looks something like

`type mongo.Client interface {
  Database(...) mongo.Database
}

type mongo.Database interface {
  Collection(...) mongo.Collection
  Drop(...) error
}

type mongo.Collection interface {
 Find()
  FindOne()
...
  DropDatabase()
  CreateDatabase()
  Drop()
}`

It sort of embed the struct hierarchy 2 levels before we can access the main function on mongo.Collection, moreover there's also TASK_DROP_DATABASE where we don't need it continue up until mongo.Collection. So I had decided to seperate it into two client (MongoDatabaseClient, MongoCollectionClient) where MongoDatabaseClient will be used only for TASK_DROP_DATABASE. I'm not sure whether it's a good implementation, please your suggestions.

Reply:
In this MongoDB case, I think it would be fine because it does not influence too much with only 2 layers.

Some engineers would like to create interface as much as possible for the future space.
But, in my opinion, I do not like over design because it increases the cost of maintenance.


Abstractly, I think coding is like creating a world and designing code structure is like grouping concept.

Let's take animal as example.
At first, the world is small, you only have dog & cat. 
So, you will only need an interface called animal.
But, when your world become bigger, there may be more kinds of cats, you may need another interface called cat for other types of cats.
However, if your world is unbalanced, there is only a dog but thousands of cat. The interface animal may not so necessary. Then, we can consider to take it off to simplify the world.
The story will be too long if I kept writing. So, I stop here.

If you are interested in this kind of concept, you can read design patterns. I feel it is like philosophy.
And, my belief is "as long as we think the code deeply with philosophy, we can increase the maintenance of codebase".

Reply:
However, I read the code here.
Probably your design makes you need to do it in main file.
If you only have a layer, you can make this logic in the taskFunction. right?
Then, you won't need to do `if e.Task != TaskDropDatabase && e.client.collectionClient == nil {` this kind of logic.

```
		var db *mongo.Database
		if e.client.databaseClient == nil {
			db = client.Database(inputStruct.DatabaseName)
			e.client.databaseClient = db
		}

		if e.Task != TaskDropDatabase && e.client.collectionClient == nil {
			collection := db.Collection(inputStruct.CollectionName)
			e.client.collectionClient = collection
			e.client.searchIndexClient = collection.SearchIndexes()
		}
```

Reply:
However, I read the code here.
Probably your design makes you need to do it in main file.
If you only have a layer, you can make this logic in the taskFunction. right?
Then, you won't need to do if e.Task != TaskDropDatabase && e.client.collectionClient == nil { this kind of logic.

```
		var db *mongo.Database
		if e.client.databaseClient == nil {
			db = client.Database(inputStruct.DatabaseName)
			e.client.databaseClient = db
		}

		if e.Task != TaskDropDatabase && e.client.collectionClient == nil {
			collection := db.Collection(inputStruct.CollectionName)
			e.client.collectionClient = collection
			e.client.searchIndexClient = collection.SearchIndexes()
		}
```

Reply:
@chuang8511

> But, when your world become bigger, there may be more kinds of cats, you may need another interface called cat for other types of cats.

Good point, in the end I've decided to keep MongoCollectionClient and MongoDatabaseClient type for component_test.go, but to group it up as to unify those two types (since both are part of mongo.Client), i've decided to make one more type

```
type MongoClient struct {
	collectionClient  MongoCollectionClient
	databaseClient    MongoDatabaseClient
	searchIndexClient MongoSearchIndexClient (This one is new, for vector search)
}
```

so I dont need to put if e.Task != TaskDropDatabase anymore and let the unused client be of nil value.

However, for both e.client.databaseClient == nil and e.client.collectionclient are needed for component_test.go since it uses mock client and preventing it to be overwritten